### PR TITLE
Route refactor

### PR DESCRIPTION
**Pull request by EdDev**
This PR is an attempt to structure the routing handling into individual steps.
Existing handling is overloading the merging operation with route entries merging, touching the interfaces and validating the routes.

The existing route integration tests have allowed a safer refactoring and have not been changed. Unit tests have been introduces with the new merge, metadata and validation changes.

Depends on #411

### CODE REVIEW

File: libnmstate/state.py [hunk 1]
```diff
@@ -64,32 +67,33 @@ def __keys(self):
...
     def __eq__(self, other):
-        return self is other or self.to_dict() == other.to_dict()
+        return self is other or self.__keys() == other.__keys()
 
-    def __ne__(self, other):
-        """
-        Workaround for python2
-        """
-        return not self.__eq__(other)
+    def __lt__(self, other):
```
Comment (cathay4t) [inline libnmstate/state.py]: Is there any real usage in nmstate on this `__lt__`  function?

Comment (cathay4t) [reply libnmstate/state.py]: OK. That's required by total_ordering.

File: libnmstate/state.py [hunk 2]
```diff
@@ -276,64 +276,41 @@ def merge_interfaces(self, other_state):
...
-        self._config_routes = merged_routes
+        absent_routes = set()
+        for routes in six.viewvalues(self.config_iface_routes):
+            for absent_route in (r for r in routes if r.absent):
+                absent_routes |= {
+                    r for r in other_routes if absent_route.match(r)
+                }
+                absent_routes |= {
+                    r for r in self_routes if absent_route.match(r)
+                }
```
Comment (cathay4t) [inline libnmstate/state.py]: The absent route should not remove routes in desire state, as user want to remove old routes and add new one.

File: libnmstate/state.py [hunk 3]
```diff
@@ -276,64 +276,41 @@ def merge_interfaces(self, other_state):
...
-
-        for route in other_state._config_routes:
-            # The other_state(running state) does not have any absent
-            # route entry and all have NEXT_HOP_INTERFACE
-            iface_name = route[Route.NEXT_HOP_INTERFACE]
-            other_iface_route_sets[iface_name].add(RouteEntry(route))
+        other_routes = set()
+        for ifname, routes in six.viewitems(other_state.config_iface_routes):
+            if self.config_iface_routes.get(ifname):
+                other_routes |= set(routes)
```
Comment (cathay4t) [inline libnmstate/state.py]: In https://github.com/nmstate/nmstate/pull/384/commits/c26b0ecd457a5431d93a8b16dee480198b35fbc1#diff-6f12cd36ee493d92b4777db9ba2780feR184 you support routes referring to non-defined interface.
But here you discard any current route to interface not found in desire state.

File: libnmstate/state.py (hunk 2 above)
Comment (EdDev) [reply libnmstate/state.py]: I will extract it to its own PR.

File: libnmstate/state.py [hunk 4]
```diff
@@ -276,64 +276,45 @@ def merge_interfaces(self, other_state):
...
-        absent_route_sets = set()
-        other_iface_route_sets = defaultdict(set)
-
-        for route in self._config_routes:
-            if route.get(Route.STATE) == Route.STATE_ABSENT:
-                absent_route_sets.add(RouteEntry(route))
+        other_routes = set()
+        for ifname, routes in six.viewitems(other_state.config_iface_routes):
+            ifstate = self.interfaces.get(ifname)
+            non_up_states = (InterfaceState.DOWN, InterfaceState.ABSENT)
```
Comment (cathay4t) [inline libnmstate/state.py]: Please move this line out of loop, there is no need to create it repeatly.

**Comment by EdDev** [comment]
@cathay4t , it is now passing all new added tests.

**Pull request merged**